### PR TITLE
Avoid instance creation if unnecessary

### PR DESCRIPTION
**Pull request by mganisin**
A cache-like mechanism introduced to DefaultClient using weakref to
avoid "memory leaks".

if instance for given entity_id exists (has other references) then it is
returned rather then creating new object representation.

### CODE REVIEW

**Comment by mganisin** [comment]
> Not sure about this - what would happend if you have 2 instances with same entity id? For example 2 instances of service on the the 2 instacnes of 3scale.

true, I'll come up with update
> 
> Also _instances will be shared also for the subclasses no? (I'm not sure about this - will chcek)

this should be safe, sublasses will have _instances but it will be _their_ attributes, no value shared. If I understood you

**Comment by pestanko** [comment]
Yep, you understood, but you are not correct. Instances **will not** have _their_ custom instance of the `_instance`

https://repl.it/@pestanko/sharinginstances

File: threescale_api/defaults.py [hunk 1]
```diff
@@ -156,7 +158,13 @@ def read(self, entity_id: int = None) -> 'DefaultResource':
         Returns(DefaultResource): Default resource
         """
         log.debug(self._log_message("[READ] Read ", entity_id=entity_id))
-        return self._instance_klass(client=self, entity_id=entity_id)
+
+        inst = self._instances.get(entity_id, lambda: None)()
```
Comment (pestanko) [inline threescale_api/defaults.py]: why is here `lambda: None`? Wouldn't be `None` enough?

**Comment by mganisin** [comment]
I will return to this idea... one day...

**Pull request closed without merging**